x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/RemoteAccessServerTransportFilter.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.security.transport;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.support.DestructiveOperations;
+import org.elasticsearch.common.util.concurrent.ThreadContext;
+import org.elasticsearch.transport.TransportRequest;
+import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.xpack.core.security.SecurityContext;
+import org.elasticsearch.xpack.core.security.authc.Authentication;
+import org.elasticsearch.xpack.security.authc.AuthenticationService;
+import org.elasticsearch.xpack.security.authc.RemoteAccessAuthenticationService;
+import org.elasticsearch.xpack.security.authz.AuthorizationService;
+
+final class RemoteAccessServerTransportFilter extends ServerTransportFilter {
+    private final RemoteAccessAuthenticationService remoteAccessAuthcService;
+
+    RemoteAccessServerTransportFilter(
+        AuthenticationService authcService,
+        AuthorizationService authzService,
+        ThreadContext threadContext,
+        boolean extractClientCert,
+        DestructiveOperations destructiveOperations,
+        SecurityContext securityContext
+    ) {
+        super(authcService, authzService, threadContext, extractClientCert, destructiveOperations, securityContext);
+        this.remoteAccessAuthcService = new RemoteAccessAuthenticationService(authcService);
+    }
+
+    @Override
+    protected void authenticate(
+        final String securityAction,
+        final TransportRequest request,
+        final ActionListener<Authentication> authenticationListener
+    ) {
+        if (securityAction.equals(TransportService.HANDSHAKE_ACTION_NAME)) {
+            super.authenticate(securityAction, request, authenticationListener);
```

Comment:
Does this work? I think the superclass will look for the regular `Authorization` header and fail to authenticate?

Reply:
This works because for the handshake action specifically, the querying cluster will send the transport handshake action with a regular authentication header, instead of remote access headers. So the super class class with look for existing authentication, and find it. 

I had a comment here before that got lost in refactoring; I'll add it back because this part is quite unclear without it. 

The code on the querying cluster side also does not make this very obvious. The way it currently works is that the transport handshake action is not included in `REMOTE_ACCESS_ACTION_ALLOWLIST` in `SecurityServerTransportInterceptor`, and `interceptForRemoteAccessRequests` decides to not use remote access credentials for any action that is not included in `REMOTE_ACCESS_ACTION_ALLOWLIST`. Especially if we re-use `REMOTE_ACCESS_ACTION_ALLOWLIST` between QC and FC code, we will probably want to make the QC-side code more explicit in handling the handshake action. I'll have a think on this 👍

Reply:
Thanks for explaining this. In that case, it should work. However, this makes me question is the behaviour on the sender side. 

1. Put handshake aside first, if we can already decide that the connection is for a new RCS remote cluster, i.e. cluster alias and credential exist, why would we fallback to the default sender? It feels unnecessarily lenient. Shouldn't we just error?
2. Now for handshake, it is rather unintuitive that this action goes through a different path than all other actions. Though techinically works, it feels like an unnecessary cognitive burden. Since we already have separate filter and authc service, I'd prefer we don't make this action an exception.

I am ok with these being a follow-up (assuming we agree with the above), just need an issue to track it.

Reply:
> Put handshake aside first, if we can already decide that the connection is for a new RCS remote cluster, i.e. cluster alias and credential exist, why would we fallback to the default sender?

This is to continue supporting CCR while CCS is enabled. Even if we have a new RCS remote cluster configured, we still want the old security model for CCR.

> Now for handshake, it is rather unintuitive that this action goes through a different path than all other actions.

Yup, this is quirky, as discussed on Slack previously. Transport handshake is historically handled separately in many places and requires `internal:transport/handshake` privilege, which we can't assign to API keys or users. As such, it can't be handled via the general approach of intersecting privileges and has to have its own "special" path regardless. We could tweak this special path to still go through `RemoteAccessAuthenticationService`, but piggy-backing on the legacy mechanism felt more pragmatic. 

Happy to explore both items in a follow up for sure. I don't like the ambiguity of `REMOTE_ACCESS_ACTION_ALLOWLIST` and want to clean that part up.

Reply:
> This is to continue supporting CCR while CCS is enabled. Even if we have a new RCS remote cluster configured, we still want the old security model for CCR.

I just realized that this part is bogus. In the current approach things will break for CCR because the QC will still send request to the new port, but it will fall back on sending the wrong headers and the FC will reject the request. I'm thinking through how this should actually work. Sorry for the confusion!

Reply:
This is my plan to address this:

1. In this PR, rename `REMOTE_ACCESS_ACTION_ALLOWLIST` to `REMOTE_CLUSTER_SECURITY_SUPPORTED_ACTION_ALLOWLIST` to clarify its purpose.
2. In this PR, add FC-side code that falls back on authenticating via the old authentication header for actions that come in on the remote access port but are not on that allowlist
3. In a follow-up PR, add the handshake action to `REMOTE_CLUSTER_SECURITY_SUPPORTED_ACTION_ALLOWLIST` (so that the QC will send it with remote access headers) and tweak the fulfilling cluster side code to handle it either inside `RemoteAccessAuthenticationService` or `RemoteAccessServerTransportFilter`, with an early if-statement checking for the handshake action and authenticating and authorizing it without intersecting with API key privileges.
4. In a follow-up PR, add a new FC-side allowlist that captures all actions allowed to come in on the new port (orchestration, CCS, and CCR).

Let me know if this makes sense.

Reply:
We had a sync around this, just summarizing the conclusions:

It's not worth the added complexity to maintain two separate allowlists (and attempting to get things to kind-of work for CCR, instead of just letting it break).

It's also better to get the end-to-end flow merged first, then iterate on improvements.

I'll work on the simplest approach to get this working for this PR, and work on follow-ups after it's merged.